Sources/Core/Sources/ECS/Systems/PlayerInputSystem.swift

```diff
@@ -146,6 +148,47 @@ public final class PlayerInputSystem: System {
         if !message.isEmpty {
           guiState.messageInput?.removeLast()
         }
+      } else if event.key == .upArrow {
+        // If no message is selected, select the above message
+        if guiState.currentMessageIndex == nil {
+          guiState.stashedMessageInput = guiState.messageInput
+          // If there are no messages, do nothing
+          if guiState.playerMessageHistory.count > 0 {
+            // Index up a message
+            guiState.currentMessageIndex = guiState.playerMessageHistory.count - 1
+            if guiState.currentMessageIndex == guiState.currentMessageIndex {
+              // Set the message input to the message at the index
+              guiState.messageInput = guiState.playerMessageHistory[guiState.currentMessageIndex!]
+            }
+          }
+        } else {
+          // If there is a message selected, index up a message
+          if guiState.currentMessageIndex! > 0 {
+            guiState.currentMessageIndex! -= 1
+          }
+          if guiState.currentMessageIndex == guiState.currentMessageIndex {
+            // Set the message input to the message at the index
+            guiState.messageInput = guiState.playerMessageHistory[guiState.currentMessageIndex!]
+          }
+        }
+      } else if event.key == .downArrow {
+        // If there is a message selected, index down a message
+        if guiState.currentMessageIndex != nil {
+          if guiState.currentMessageIndex! < guiState.playerMessageHistory.count - 1 {
+            guiState.currentMessageIndex! += 1
+            if guiState.currentMessageIndex == guiState.currentMessageIndex {
+              guiState.messageInput = guiState.playerMessageHistory[guiState.currentMessageIndex!]
+            }
+          } else {
+            // If there is no message to index down to, go back to what the user was typing originally
+            guiState.currentMessageIndex = nil
+            if guiState.stashedMessageInput != nil {
+              guiState.messageInput = guiState.stashedMessageInput
+            } else {
+              guiState.messageInput = ""
```

Comment:
This would clear the input if a user presses the down arrow while typing a message if they haven’t pressed the up arrow yet. I don’t believe this is the behaviour in vanilla

Reply:
Never mind